st2api/st2api/controllers/v1/auth.py

```diff
@@ -0,0 +1,171 @@
+    @jsexpose(arg_types=[str])
+    def get_one(self, api_key_id_or_key):
+        """
+            List api keys.
+
+            Handle:
+                GET /apikeys/1
+        """
+        api_key_db = None
+        try:
+            api_key_db = ApiKey.get_by_key_or_id(api_key_id_or_key)
+        except ApiKeyNotFoundError:
+            msg = 'ApiKey matching %s for reference and id not found.', api_key_id_or_key
+            LOG.exception(msg)
+            abort(http_client.NOT_FOUND, msg)
+
+        try:
+            return ApiKeyAPI.from_model(api_key_db, mask_secrets=True)
+        except (ValidationError, ValueError) as e:
+            LOG.exception('Failed to serialize API key.')
+            abort(http_client.INTERNAL_SERVER_ERROR, str(e))
+
+    @request_user_has_permission(permission_type=PermissionType.API_KEY_VIEW)
+    @jsexpose(arg_types=[str])
+    def get_all(self, **kw):
+        """
+            List all keys.
+
+            Handles requests:
+                GET /keys/
+        """
+
+        api_key_dbs = ApiKey.get_all(**kw)
+        api_keys = [ApiKeyAPI.from_model(api_key_db, mask_secrets=True)
+                    for api_key_db in api_key_dbs]
+
+        return api_keys
+
+    @request_user_has_permission(permission_type=PermissionType.API_KEY_CREATE)
+    @jsexpose(body_cls=ApiKeyAPI, status_code=http_client.CREATED)
+    def post(self, api_key_api):
+        """
+        Create a new entry or update an existing one.
+        """
+        api_key_db = None
+        try:
+            api_key_api.user = self._get_user(api_key_api)
+            api_key, api_key_hash = auth_util.generate_api_key_and_hash()
+            # store key_hash in DB
+            api_key_api.key = api_key_hash
+            api_key_db = ApiKey.add_or_update(ApiKeyAPI.to_model(api_key_api))
+        except (ValidationError, ValueError) as e:
+            LOG.exception('Validation failed for api_key data=%s.', api_key_api)
+            abort(http_client.BAD_REQUEST, str(e))
+
+        extra = {'api_key_db': api_key_db}
+        LOG.audit('ApiKey created. ApiKey.id=%s' % (api_key_db.id), extra=extra)
+
+        api_key_api = ApiKeyAPI.from_model(api_key_db)
+        # Return real api_key back to user. A one-way hash of the api_key is stored in the DB
+        # only the real value only returned at create time. Also, no masking of key here since
+        # the user needs to see this value atleast once.
+        api_key_api.key = api_key
+        return api_key_api
+
+    @request_user_has_resource_permission(permission_type=PermissionType.API_KEY_DELETE)
+    @jsexpose(arg_types=[str], status_code=http_client.NO_CONTENT)
+    def delete(self, api_key_id_or_key):
+        """
+            Delete the key value pair.
+
+            Handles requests:
+                DELETE /apikeys/1
+        """
+        api_key_db = None
+        try:
+            api_key_db = ApiKey.get_by_key_or_id(api_key_id_or_key)
```

Comment:
You can let "object not found" exceptions to propagate.

They are caught in the response hook where they are also translated to the corresponding error message and status code.